# Planet API Python Client


This tutorial is an introduction to [Planet](https://www.planet.com)'s Data API using the official [Python client](https://github.com/planetlabs/planet-client-python), the `planet` module.

## Requirements

This tutorial assumes familiarity with the [Python](https://python.org) programming language throughout. Python modules used in this tutorial are:
* [IPython](https://ipython.org/) and [Jupyter](https://jupyter.org/)
* [planet](https://github.com/planetlabs/planet-client-python)
* [geojsonio](https://pypi.python.org/pypi/geojsonio)

You should also have an account on the Planet Platform and retrieve your API key from your [account page](https://www.planet.com/account/).

## Useful links 
* [Documentation](https://planetlabs.github.io/planet-client-python/index.html)
* [Planet Data API reference](https://www.planet.com/docs/reference/data-api/)

This tutorial will cover the basic operations possible with the Python client, particularly those that interact with the Data API.

The basic workflow for interacting with the Data API is:
1. search item types based on filters
1. activate assets
1. download assets

## Set up

In order to interact with the Planet API using the client, we need to import `planet.api`.

In [1]:
from planet import api

We next need to create a `ClientV1` object registered with our API key. The API key will be automatically read from the `PL_API_KEY` environment variable if it exists. If not it can be provided when creating the `api` object.

In [2]:
client = api.ClientV1()
# client = api.ClientV1(api_key="abcdef0123456789") <-- not a real key

# print client.auth.value

`ClientV1` provides basic low-level access to Planet’s API. Only one `ClientV1` should be in existence for an application. The client is thread safe and takes care to avoid API throttling and also retries any throttled requests. Any exceptional HTTP responses are handled by translation to one of the `planet.api.exceptions` classes.

We will also create a small helper function to print out JSON with proper indentation.

In [3]:
import json

def p(data):
    print(json.dumps(data, indent=2))

Let's also read in a GeoJSON geometry into a variable so we can use it during testing.

In [4]:
with open("data/san-francisco.json") as f:
    geom = json.loads(f.read())

## Searching

We can search for items that are interesting by using the `quick_search` member function. Searches, however, always require a proper request that includes a filter that selects the specific items to return as seach results.

### Filters

The Planet Python client also includes `planet.api.filters` to assist in the creation of search filters.

In [5]:
from planet.api import filters

The possible filters include `and_filter`, `date_filter`, `range_filter` and so on, mirroring the options supported by the Planet API.


In [6]:
from datetime import datetime
start_date = datetime(year=2017, month=1, day=1)

date_filter = filters.date_range('acquired', gte=start_date)
cloud_filter = filters.range_filter('cloud_cover', lte=0.1)

In [7]:
and_filter = filters.and_filter(date_filter, cloud_filter)

In [8]:
p(and_filter)

{
  "type": "AndFilter",
  "config": [
    {
      "field_name": "acquired",
      "type": "DateRangeFilter",
      "config": {
        "gte": "2017-01-01T00:00:00Z"
      }
    },
    {
      "field_name": "cloud_cover",
      "type": "RangeFilter",
      "config": {
        "lte": 0.1
      }
    }
  ]
}


In addition to the filter, a properly-constructed request also contains the list of item types that we want to select.

In [9]:
item_types = ["REOrthoTile", "PSOrthoTile"]
req = filters.build_search_request(and_filter, item_types)

In [10]:
p(req)

{
  "item_types": [
    "PSOrthoTile",
    "REOrthoTile"
  ],
  "filter": {
    "type": "AndFilter",
    "config": [
      {
        "field_name": "acquired",
        "type": "DateRangeFilter",
        "config": {
          "gte": "2017-01-01T00:00:00Z"
        }
      },
      {
        "field_name": "cloud_cover",
        "type": "RangeFilter",
        "config": {
          "lte": 0.1
        }
      }
    ]
  }
}


In [11]:
res = client.quick_search(req)

The results of `quick_search` can be handled in different ways, but most commonly the user will either iterating through the list of items (`items_iter`) or writing items to a GeoJSON file (`json_encode`).

In either case, the number of items must be specified.

In [12]:
for item in res.items_iter(4):
    print(item['id'], item['properties']['item_type'])

5524910_4456218_2022-03-28_2414 PSOrthoTile
5524910_4452611_2022-03-28_2414 PSOrthoTile
5524910_4451208_2022-03-28_2414 PSOrthoTile
5524910_4452311_2022-03-28_2414 PSOrthoTile


If the number of items requested is more than 250, the client will automatically fetch more pages of results in order to get the exact number requested.

In [13]:
with open('output/results.json','w') as f:
    res.json_encode(f,1000)

This GeoJSON file can be opened and viewed in any compatible application.

## Assets and downloads

After a search returns results, the Python client can be used to check for assets and initiate downloads.

The list of assets for an item can be retrieved with `get_assets` or `get_assets_by_id`.

In [14]:
print(item['id'])

5524910_4452311_2022-03-28_2414


In [22]:
assets = client.get_assets(item).get()

In [23]:
assets

{'analytic': {'_links': {'_self': 'https://api.planet.com/data/v1/assets/eyJpIjogIjU1MjQ5MTBfNDQ1MjMxMV8yMDIyLTAzLTI4XzI0MTQiLCAiYyI6ICJQU09ydGhvVGlsZSIsICJ0IjogImFuYWx5dGljIiwgImN0IjogIml0ZW0tdHlwZSJ9',
   'activate': 'https://api.planet.com/data/v1/assets/eyJpIjogIjU1MjQ5MTBfNDQ1MjMxMV8yMDIyLTAzLTI4XzI0MTQiLCAiYyI6ICJQU09ydGhvVGlsZSIsICJ0IjogImFuYWx5dGljIiwgImN0IjogIml0ZW0tdHlwZSJ9/activate',
   'type': 'https://api.planet.com/data/v1/asset-types/analytic'},
  '_permissions': ['download'],
  'md5_digest': None,
  'status': 'inactive',
  'type': 'analytic'},
 'analytic_5b': {'_links': {'_self': 'https://api.planet.com/data/v1/assets/eyJpIjogIjU1MjQ5MTBfNDQ1MjMxMV8yMDIyLTAzLTI4XzI0MTQiLCAiYyI6ICJQU09ydGhvVGlsZSIsICJ0IjogImFuYWx5dGljXzViIiwgImN0IjogIml0ZW0tdHlwZSJ9',
   'activate': 'https://api.planet.com/data/v1/assets/eyJpIjogIjU1MjQ5MTBfNDQ1MjMxMV8yMDIyLTAzLTI4XzI0MTQiLCAiYyI6ICJQU09ydGhvVGlsZSIsICJ0IjogImFuYWx5dGljXzViIiwgImN0IjogIml0ZW0tdHlwZSJ9/activate',
   'type': 'https://api.p

In [25]:
for asset in sorted(assets.keys()):
    print(asset)

analytic
analytic_5b
analytic_5b_xml
analytic_dn
analytic_dn_xml
analytic_xml
udm
udm2
visual
visual_xml


In [26]:
activation = client.activate(assets['analytic'])
activation.response.status_code

202

A response of 202 means that the request has been accepted and the activation will begin shortly. A 204 code indicates that the asset is already active and no further action is needed. A 401 code means the user does not have permissions to download this file.

Below, we are polling the API until the item is done activation. This may take awhile.

In [27]:
import time

asset_activated = False

while asset_activated == False:
    
    # Get asset and its activation status
    assets = client.get_assets(item).get()
    asset = assets.get('analytic')
    asset_status = asset["status"]
    
    # If asset is already active, we are done
    if asset_status == 'active':
        asset_activated = True
        print("Asset is active and ready to download")
   
    # Still activating. Wait and check again.
    else:
        print("...Still waiting for asset activation...")
        time.sleep(3)

...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting

...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting

...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
...Still waiting for asset activation...
Asset is active and ready to download


In [33]:
callback = api.write_to_file(directory='output/')
body = client.download(assets['analytic_xml'], callback=callback)
body.await()

SyntaxError: invalid syntax (<ipython-input-33-82ebe149417f>, line 3)

## Saved Searches

The Python API client can also help in managing saved searches on the Planet Platform.

In [34]:
searches = client.get_searches()

In [35]:
for search in searches.items_iter(100):
    print(search['id'], search['name'])

40bfde64d0534166b6f7e8c7fc8c89fd Vancouver Island
950f543836a2459e9c4d5848c8219304 Raj Ghat, New Delhi, Central Delhi, Delhi, India
ae24a8d7251244b1ac283bc5b9398b78 San Francisco
a7a648fca59245bca42d2466b289f34f San Francisco
6950887354d84e23889c3e308355fdcf Vancouver Island
fe44f67590ae44d6a4e5e8030de85f90 Vancouver Island


In [36]:
item_types = ["PSScene"]
san_francisco_filter = filters.geom_filter(geom)
req = filters.build_search_request(san_francisco_filter, item_types, name="San Francisco")

In [37]:
p(req)

{
  "item_types": [
    "PSScene"
  ],
  "filter": {
    "field_name": "geometry",
    "type": "GeometryFilter",
    "config": {
      "type": "Polygon",
      "coordinates": [
        [
          [
            -122.47455596923828,
            37.810326435534755
          ],
          [
            -122.49172210693358,
            37.795406713958236
          ],
          [
            -122.52056121826172,
            37.784282779035216
          ],
          [
            -122.51953124999999,
            37.6971326434885
          ],
          [
            -122.38941192626953,
            37.69441603823106
          ],
          [
            -122.38872528076173,
            37.705010235842614
          ],
          [
            -122.36228942871092,
            37.70935613533687
          ],
          [
            -122.34992980957031,
            37.727280276860036
          ],
          [
            -122.37773895263672,
            37.76230130281876
          ],
          [
     

In [38]:
res = client.create_search(req)

In [39]:
search = res.get()
print(search["id"], search["name"])

2c2d202cd5a343699e4b1f677e97ed8d San Francisco


In [40]:
res = client.saved_search(search["id"])

In [41]:
for item in res.items_iter(20):
    print(item["id"], item["properties"]["view_angle"])

20220326_180313_91_241f 3
20220326_180311_62_241f 3
20220327_183227_22_2490 1
20220327_183229_53_2490 1
20220327_180339_22_2435 5
20220327_180341_52_2435 5
20220327_180456_67_106c 5
20220327_180455_17_106c 5
20220327_180453_67_106c 5
20220325_175739_95_2455 3.9
20220325_175737_65_2455 4
20220324_180251_82_2430 2.1
20220324_180249_35_2430 2.1
20220323_183536_13_2461 5
20220323_183538_46_2461 5
20220323_180500_58_106a 3
20220324_182935_1009 0.1
20220324_182934_1009 0.1
20220324_182933_1009 0.1
20220322_183117_0f4e 3.4


# Statistics

The Python API client can also help report statistical summaries of the amount of data in the Planet API.

In [42]:
item_types = ["PSScene"]
san_francisco_filter = filters.geom_filter(geom)
req = filters.build_search_request(san_francisco_filter, item_types, interval="year")

In [43]:
stats = client.stats(req).get()

In [44]:
p(stats)

{
  "buckets": [
    {
      "count": 3,
      "start_time": "2014-01-01T00:00:00.000000Z"
    },
    {
      "count": 60,
      "start_time": "2015-01-01T00:00:00.000000Z"
    },
    {
      "count": 200,
      "start_time": "2016-01-01T00:00:00.000000Z"
    },
    {
      "count": 1076,
      "start_time": "2017-01-01T00:00:00.000000Z"
    },
    {
      "count": 1674,
      "start_time": "2018-01-01T00:00:00.000000Z"
    },
    {
      "count": 1452,
      "start_time": "2019-01-01T00:00:00.000000Z"
    },
    {
      "count": 1514,
      "start_time": "2020-01-01T00:00:00.000000Z"
    },
    {
      "count": 1617,
      "start_time": "2021-01-01T00:00:00.000000Z"
    },
    {
      "count": 292,
      "start_time": "2022-01-01T00:00:00.000000Z"
    }
  ],
  "interval": "year",
  "utc_offset": "+0h"
}


In [45]:
assets = client.get_assets(item)

In [46]:
assets.last_modified()

In [47]:
assets.get()

{'basic_analytic_4b': {'_links': {'_self': 'https://api.planet.com/data/v1/assets/eyJpIjogIjIwMjIwMzIyXzE4MzExN18wZjRlIiwgImMiOiAiUFNTY2VuZSIsICJ0IjogImJhc2ljX2FuYWx5dGljXzRiIiwgImN0IjogIml0ZW0tdHlwZSJ9',
   'activate': 'https://api.planet.com/data/v1/assets/eyJpIjogIjIwMjIwMzIyXzE4MzExN18wZjRlIiwgImMiOiAiUFNTY2VuZSIsICJ0IjogImJhc2ljX2FuYWx5dGljXzRiIiwgImN0IjogIml0ZW0tdHlwZSJ9/activate',
   'type': 'https://api.planet.com/data/v1/asset-types/basic_analytic_4b'},
  '_permissions': ['download'],
  'expires_at': '2022-03-28T15:34:16.113560',
  'location': 'https://api.planet.com/data/v1/download?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJzdWIiOiJEc1NwV09nMHl4WnZ6Y0piSFFnd3c1NHlTTTJiM0h4ZEluakFGekhMWHhGaE1nR2hWUFJOQzNkMHg2WlRVcEx0U1lEYjZUSnc0NDVFalRSUXZXRHRDUT09IiwiZXhwIjoxNjQ4NDgxNjU2LCJ0b2tlbl90eXBlIjoidHlwZWQtaXRlbSIsIml0ZW1fdHlwZV9pZCI6IlBTU2NlbmUiLCJpdGVtX2lkIjoiMjAyMjAzMjJfMTgzMTE3XzBmNGUiLCJhc3NldF90eXBlIjoiYmFzaWNfYW5hbHl0aWNfNGIifQ.tXn20QXUF6ybbFUxsH3JH_wz3ozuCo3RLmpwveRWafbe4u